### ISSUE TITLE
auto-create topics on server not supported by library.

### ISSUE THREAD

**Issue by butlermatt** [pulyaevskiy/dart-kafka#5]
Currently the library doesn't support the option on a server to auto-create topics.

A broker can be configured to auto-create a topics if it is published to when it does not exist. However currently the library validates that the topic is found in the meta data and throws an error if it is not found.

**Comment by pulyaevskiy** [comment]
Thanks!

According to Kafka [documentation](https://kafka.apache.org/08/configuration.html) fetching metadata for particular topic should also auto create it.

You can try to temporarily workaround it by explicitly specifying topic name in `KafkaSession.getMetadata`, e.g.:

``` dart
var session = new KafkaSession(new ContactPoint('host', 9092));
var metadata = await session.getMetadata(topicNames: ['your_topic_name']);
// topic should be created at this point so consume and publish should work.
```

I also have to warn you that current implementation of metadata cache in `KafkaSession` may not handle this use case properly. 

I might have to refactor it to not return `MetadataResponse` but some aggregated value object instead.

Will submit a PR for this issue asap.

Thanks again.

**Comment by butlermatt** [comment]
Using the metadata trick above only half works. It does create the topic on the server, but the library still reports an error, in this case:

```
Bad state: No element
#0      Object&ListMixin.firstWhere (dart:collection/list.dart:133)
#1      TopicMetadata.getPartition (package:kafka/src/protocol/metadata_api.dart:97:18)
#2      Producer.produce.<produce_async_body> (package:kafka/src/producer.dart:42:29)
#3      _RootZone.runUnary (dart:async/zone.dart:1149)
#4      _Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:502)
#5      _Future._propagateToListeners (dart:async/future_impl.dart:585)
#6      _Future._complete (dart:async/future_impl.dart:366)
#7      _SyncCompleter.complete (dart:async/future_impl.dart:52)
#8      KafkaSession.getMetadata.<getMetadata_async_body> (package:kafka/src/session.dart)
#9      Future.Future.microtask.<anonymous closure> (dart:async/future.dart:144)
#10     _microtaskLoop (dart:async/schedule_microtask.dart:43)
#11     _microtaskLoopEntry (dart:async/schedule_microtask.dart:52)
#12     _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:96)
#13     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:151)
```

But it only seems to happen the first time. If I then call the same method, either produce or a consume, then it will work as expected. It seems there needs to be some type of additional delay beyond the await for the `session.getMetadata` for a different future to complete before the cache is updated.

Even after introducing an async delay before retrying a consume/produce I still get the bad state until I make a second call.

**Comment by pulyaevskiy** [comment]
Yeah, that's why I mentioned it may not handle this case properly.

I'm working on a fix already and will submit a PR soon.

Thanks again for your input!

**Comment by pulyaevskiy** [comment]
@butlermatt I've just merged #9 . It introduces new `ClusterMetadata` and should work with auto-created topics.

I'm closing this one but feel free to re-open if you find any issues!
